design/0003-byop.md

```diff
@@ -0,0 +1,158 @@
+
+## Using Exisiting Prometheus
+
+First, Documentation will be provided at `linkerd.io` to guide users on
+adding the `scrape-configs` required, to get the metrics from linkerd
+proxies, control plane components, etc. This can be found
+[here](https://github.com/linkerd/linkerd2/blob/master/charts/linkerd2/templates/prometheus.yaml#L26)
+
+Once this is done, Users should be able to see golden metrics, etc in their
+exisiting prometheus instance. The `prometheus.url` field helps users in
+configuring Linkerd components i.e `public-api`,etc that needs prometheus
+access to the exisiting prometheus installation.
+
+Flag interactions are defined below:
+
+| `prometheus.enabled` | `prometheus.url`              | Result                                                                                                                                                                                                                                           |
+|---------|------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| true    | ""               | Install linkerd-prometheus, and configure control-plane components to use it.                                                                                                                                                                    |
+| false   | "prometheus.xyz" | No linkerd-prometheus, control-plane components will be configured to url                                                                                                                                                                        |
+| true    | "prometheus.xyz" | Install linkerd-prometheus, but stil configure control plane to use different url, useful when linkerd-prometheus is configured to remote-write into long term storage solution at url, Thus  powering dashboard and CLI through long term storage. |
+| false   | ""               | No linkerd-prometheus, and no URL to power dashboards. (Should we even allow this?)                                                                                                                                                              |
+
+### Linkerd Check Tooling
+
+When users get their own Prometheus, It would be really great to have
+some tooling that allows users to check if `scrape-configs` are configured
+correctly and prometheus is able to get metrics.
+
+`--prometheus` flag can be added to linkerd check to run these tests.
+The following tests are planned:
+
+- Check if the required scrape-configs are present in proemetheus cm.
+( It should be noted that, the prometheus could be present in an differents
+namespace, etc)
+
+- Check if the common metrics are queriable in prometheus, by using the
+end-point i.e `prometheus.url` from the `linkerd-values` cm.
```

Comment:
A downside of trying to check the configmap is that we might not know which one it is, nor have access to it.

A downside of querying prometheus is that that might not give us enough info to verify the latest scrape config has been adopted (some subtle scrape config changes might not be reflected easily as a query result). I wonder if we could add some sort of version number in the scrape config that could be reflected as a query result...

Reply:
> A downside of trying to check the configmap is that we might not know which one it is, nor have access to it.

Hmmm, we can mitigate this, by making the user pass the configmap name to the CLI, but there will still be the access problem but the good side should be that we can also have a command which tires to update the scrape config to the current version's scrape config.

> I wonder if we could add some sort of version number in the scrape config that could be reflected as a query result...

This is definitely interesting. but from what I searched, I couldn't find a way to add a param into the scrape-config but can we have a metric on the scrape-config version? If yes who would be the right candidate to check and add for the metric? Also, wondering if there would be any inconsistencies on the actual config, and this metric.